CRM-21316: Add missing code to handle smart group on sms mode

monishdeb opened the pull request:
Overview
----------------------------------------
This is a break on ```CRM_Mailing_BAO_Mailing::getRecipients()``` where during refactoring, the code responsible to fetch contacts of selected smart group IDs, on SMS mode was missed. And this PR is about retaining that code and some additional changes. Thanks to @JKingsnorth for notifying this break.

---

 * [CRM-21316: Refactor getRecipients fn by creating a new fn getMailRecipients, that will be used ONLY to fetch and store mail recipients](https://issues.civicrm.org/jira/browse/CRM-21316)

monishdeb commented:
@JKingsnorth can you please review my patch?

eileenmcnaughton commented on CRM/Mailing/BAO/Mailing.php:
```diff
@@ -218,30 +219,31 @@ public static function getRecipients($mailingID) {
+        "$contact.do_not_sms",
+        "$contact.is_opt_out = 0",
+        "$contact.is_deceased <> 1",
+        "$phone.phone_type_id = " . CRM_Utils_Array::value('Mobile', CRM_Core_OptionGroup::values('phone_type', TRUE, FALSE, FALSE, NULL, 'name')),
```
does CRM_Core_PseudoConstant::getKey not work here? It's certainly shorter & that OptionGroup::values function is a bit evil

eileenmcnaughton commented:
UT would be great here.

JKingsnorth commented on CRM/Mailing/BAO/Mailing.php:
```diff
@@ -260,21 +262,38 @@ public static function getRecipients($mailingID) {
+          ->param('#groups', $includeSmartGroupIDs)
+          ->param('#mailingID', $mailingID)
+          ->execute();
+      }
```
This is almost all duplicated - could we just store the label of the ID field in a variable to use? This should help avoid SMS being accidentally omitted if anything else were to change. Or do you think that would make it unnecessarily hard to read?

JKingsnorth commented on CRM/Mailing/BAO/Mailing.php:
```diff
@@ -218,30 +219,31 @@ public static function getRecipients($mailingID) {
+        'temp.contact_id IS null',
+      );
       CRM_Utils_SQL_Select::from($phone)
         ->select(" DISTINCT $phone.id as phone_id ")
```
@monishdeb This needs to also select contact id, otherwise it causes a fatal error because the column count doesn't match up with replace into.

JKingsnorth commented:
@monishdeb Additionally, we can't rely on this here: https://github.com/JMAConsulting/civicrm-core/blob/038b55fd5b9407b4f10eb9666c0106a24173cc9b/CRM/Mailing/BAO/Mailing.php#L126

Because the SMS provider is set on the second step of creating an SMS. So the first time this function is called, the SMS provider ID is not set, and it tries to calculate email recipients for the SMS.

We could fix this by passing a 'dummy' SMS provider ID from the first page of the 'New SMS' form perhaps?

JKingsnorth commented:
I have started on some tests for the getRecipients function as part of https://github.com/civicrm/civicrm-core/pull/11162 . I'd like to set aside some time to write some further tests for this.

monishdeb commented:
@JKingsnorth I am not getting enough time (working on an internal project, need to be deployed next week) to make the additional changes as required in this PR. Can you please submit a patch on top my PR changes. And I will include that patch as commit of yours, in this PR. Or if it's not urgent can I submit the final changes by next week itself?

@michaelmcandrew I saw your PR https://github.com/civicrm/civicrm-core/pull/11589, can I include your fix this PR?

michaelmcandrew commented:
@monishdeb - sure :)

michaelmcandrew commented:
@JKingsnorth - since you are interested in SMS tests, I wanted to point you in the direction of @mattwire's [open pr with some sms tests](https://github.com/civicrm/civicrm-core/pull/10946) in case you hadn't seen it. (I had forgotten about it until today).

monishdeb commented:
First of all, sorry for delay :( I have updated the PR and here are new changes:
1. Cherry-picked @michaelmcandrew commit for https://github.com/civicrm/civicrm-core/pull/11558
2. Cherry-picked @JKingsnorth commit to include SMS unit test https://github.com/civicrm/civicrm-core/pull/11558/commits/31a5c38f402c639b8e8630a3255b5b714558399b 
3. Refactored code to use same SQL queries for both mailing and sms
4. Modified unit-test to use smart group and other style fixes.

@eileenmcnaughton @JKingsnorth @michaelmcandrew thankyou all for your feedback and fixes made for this regression. Can you please review my latest patch so that we can merge it against 4.7.31-rc ?

seamuslee001 commented:
Jenkins re test this pease

seamuslee001 commented:
Jenkins re test this please

seamuslee001 commented on CRM/Mailing/BAO/Mailing.php:
```diff
@@ -201,79 +203,71 @@ public static function getRecipients($mailingID) {
+        ->join('temp', " LEFT JOIN $excludeTempTablename temp ON $contact.id = temp.contact_id ")
         ->where('gc.group_id IN (#groups)')
         ->where($includeFilters)
-        ->orderBy($order_by)
```
we don't need the orderby anymore? @monishdeb

eileenmcnaughton commented:
Just a note that I have not been following this closely but I know others have so I'm happy to merge if @seamuslee001 @JKingsnorth or @michaelmcandrew approve it

JKingsnorth commented:
I'll try to take a look at this early on Monday. Thanks for merging in the test and fix for the SMS ordering.

eileenmcnaughton commented:
test this please

JKingsnorth commented:
Looking at this one now. I'm running through some manual tests, but once this issue is merged I'm going to work on some more unit tests for getRecipients.

JKingsnorth commented:
@monishdeb @michaelmcandrew  This still isn't working at all for SMS mailings. The count is incorrect on the second step of sending a mass SMS. The code:
`$isSMSmode = (!CRM_Utils_System::isNull($mailingObj->sms_provider_id));`

at the beginning of getRecipients doesn't work because an sms_provider_id is not set during the first step of creating an SMS mailing. This means that getRecipients treats it as an 'email', and bases the recipient list off of whether they have an email address.

This is currently not covered by the test because the test creates the mailing with the SMS provider ID.

We can resolve this by setting the SMS provider ID to the default SMS provider on this step of the form - like: https://github.com/civicrm/civicrm-core/pull/11628

(As a side note - you can have TWO default SMS providers (separate issue), hence the way I've structured the API call...)

This seems to get things working; but I'm going to keep testing the SMS inclusion/exclusion for a few minutes.

JKingsnorth commented:
@monishdeb Additionally, the functionality to include/exclude **recipients of previous mailings** does not appear to work for email or SMS mailings?

JKingsnorth commented:
(also came cross this whilst testing, but it's a separate issue: https://issues.civicrm.org/jira/browse/CRM-21736)

monishdeb commented:
Thanks @JKingsnorth for your feedback let me cover each of your points:
> 1.  This still isn't working at all for SMS mailings. The count is incorrect on the second step of sending a mass SMS. The code:
$isSMSmode = (!CRM_Utils_System::isNull($mailingObj->sms_provider_id));

This would be a separate issue as the current patch is only about retaining the earlier code which got missed during refactoring. 

> Additionally, the functionality to include/exclude recipients of previous mailings does not appear to work for email or SMS mailings?

Yes, it should because the same table is used to fetch the exclude and include group ids [here](https://github.com/JMAConsulting/civicrm-core/blob/72c882eb74ba6329265998880777a5157b821fd4/CRM/Mailing/BAO/Mailing.php#L136) and eventually same SQL queries is used filter out contacts which are present in exclude groups.

JKingsnorth commented:
I think these issues are within scope for this, since they were all introduced as a side-effect of https://github.com/civicrm/civicrm-core/pull/11142. Reverting that PR fixes all these issues as far as I can tell.

**Cannot include 'previous recipients' in mailing**
There doesn't seem to be anything in the new SQL that includes the groups used in previous mailings? If you try to include a 'previous mailing' then the estimated recipients does not change. I am not sure whether this is due to a UI issue, or a functional issue (see below). Reverting 11142 makes this work again. 

**UI is buggy when adding/removing recipient groups**
- Add a group to a mailing
- Click the x to remove it - it is not removed from the UI

- Add a 'previous mailing' recipient list to a mailing
- Note that it does not appear
- Add a normal group
- Now they both appear

**SMS recipient estimation doesn't work on second step of creating SMS**
This is actually caused by the original PR too. In the original code we passed in a $mode to the getRecipients function, so the 'isSMS' detection worked correctly. That parameter was removed as part of the original PR, so SMS mode detection stopped working. I've suggested a workaround, as per https://github.com/civicrm/civicrm-core/pull/11628

If we have time to fix these in this release cycle then great, I can help with the PRs. Otherwise I think we should revert https://github.com/civicrm/civicrm-core/pull/11142 until we have unit tests in place to ensure continuity.

There have undoubtedly been performance improvements thanks to @monishdeb 's hard work on this area. But I don't think we can introduce this many regressions in such a sensitive area of the system.

eileenmcnaughton commented:
@totten @colemanw @mlutfy adding you in on this as we may need to make a decision to revert & try again next release if this can't be sorted out in time. Current status is that a patch merged to master (in time for 4.7.31) is causing regression/s that Monish & John are working on. Monish's time has opened up a bit so we may be fine but you should be monitoring this

monishdeb commented:
I have fixed the 'UI is buggy when adding/removing recipient groups' issue. Going to fix the 'previous mailing' issue shortly and will update the PR.

monishdeb commented:
@eileenmcnaughton @JKingsnorth I have fixed the two regressions where earlier:
1. If we select any prior mailing in recipient field then it doesn't fetch the prior recipients from ```civicrm_mailing_recipients``` table. 
As per refactoring, I have retained the query that handles this scenario.

2. We cannot unselect the groups/prior mailing from recipients field.
This was a regression due to a bug fix where earlier there was a one-line change made in Recipients.js to fix an API error when no group/mailing was selected.

Here's the screencast, after the fix:
![test-multiple-after](https://user-images.githubusercontent.com/3735621/35842657-87a0f87a-0b29-11e8-82b2-24e2fbbb4098.gif)

eileenmcnaughton commented:
Well done @monishdeb - @JKingsnorth await your review on this to merge

JKingsnorth commented:
Cool, thanks @monishdeb, testing this now.

JKingsnorth commented:
Yes, thanks @monishdeb, confirmed this version fixes:
- The UI issues
- Including/excluding previous mailing recipients

I'm working on the PR to fix SMS recipient estimation now. Just a couple of things I want to test/investigate there; but we can do that separately.

monishdeb commented:
Great. Thanks a lot for your persistent response and getting all the regression fixed :) 

Can we merge this PR ?

JKingsnorth commented:
Thumbs up from me to merge.

eileenmcnaughton commented:
merged based on @JKingsnorth review

JKingsnorth commented:
Fastest merging finger in the Southern Hemisphere @eileenmcnaughton .

eileenmcnaughton commented:
@JKingsnorth @monishdeb @michaelmcandrew is this closable now? https://issues.civicrm.org/jira/browse/CRM-21713

eileenmcnaughton commented:
quickdraw-mcGraw

JKingsnorth commented:
@eileenmcnaughton yes indeed

monishdeb commented:
Yup can be closed.

JKingsnorth commented:
I'm starting work on a suite of tests for getRecipients, https://issues.civicrm.org/jira/browse/CRM-21739

JKingsnorth commented:
And the last fix in the other PR has passed the test suite: https://github.com/civicrm/civicrm-core/pull/11628
